### PR TITLE
Ifpack2: Add Schur complements to the BlockTriDiag solver.

### PR DESCRIPTION
**Pull request by kliegeois**
@trilinos/ifpack2 

This PR adds a Schur complement approach to the BlockTriDiag solver to increase the amount of work that can be done in parallel.

### CODE REVIEW

**Review by lucbv — CHANGES_REQUESTED** [review]
Overall looks good but need to address some points

File: packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp [hunk 1]
```diff
@@ -1208,79 +924,321 @@ namespace Ifpack2 {
...
+                partptr_sub(sub_ip, 0) = partptr_sub(nparts*2*(n_subparts_per_part-1) + ip - 1, 1);
+              }
+              partptr_sub(sub_ip, 1) = sub_line_length + partptr_sub(sub_ip, 0);
+              partptr_sub(schur_ip, 0) = partptr_sub(sub_ip, 1);
+              partptr_sub(schur_ip, 1) = connection_length + partptr_sub(schur_ip, 0);
+
+              part2rowidx0_sub(sub_ip + 1) = part2rowidx0_sub(sub_ip) + sub_line_length;
+              part2rowidx0_sub(sub_ip + 2) = part2rowidx0_sub(sub_ip + 1) + connection_length;
+
+              //printf("Sub Part index = %d, first LID associated to the sub part = %d, sub part size = %d;\n", sub_ip, partptr_sub(ip, 2 * local_sub_ip), sub_line_length);
```
[the remark this answers is not in the document]
Comment (kliegeois) [reply packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp:1024]: Do you suggest adding a `member.team_barrier();` here?
There were no barriers in the original BTD code so I did not introduce one here so I assumed that this was not required.

File: packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp [hunk 2]
```diff
@@ -1208,79 +924,321 @@ namespace Ifpack2 {
...
+              interf.row_contiguous = false;
+          }
+          partptr(ip+1) = os + ipnrows;
+        }
+        part2rowidx0_sub(0) = 0;
+        partptr_sub(0, 0) = 0;
+
+        for (local_ordinal_type ip=0;ip<nparts;++ip) {
+          const local_ordinal_type ipnrows = 1;
+          //const local_ordinal_type first_sub_part_index = ip * (2*n_subparts_per_part - 1);
```
[the remark this answers is not in the document]
Comment (kliegeois) [reply packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp:988]: There is no reason; I removed the `deep_copy`.

File: packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp [hunk 3]
```diff
@@ -1208,79 +924,321 @@ namespace Ifpack2 {
...
+            if (interf.row_contiguous && os+i > 0 && lclrow((os+i)-1) + 1 != lcl_row)
+              interf.row_contiguous = false;
+          }
+          partptr(ip+1) = os + ipnrows;
+        }
+        part2rowidx0_sub(0) = 0;
+        partptr_sub(0, 0) = 0;
+
+        for (local_ordinal_type ip=0;ip<nparts;++ip) {
+          const local_ordinal_type ipnrows = 1;
```
[the remark this answers is not in the document]
Comment (lucbv) [reply packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp]: Seems odd, what did the compiler say?

File: packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp [hunk 4]
```diff
@@ -1290,26 +1247,65 @@ namespace Ifpack2 {
...
+
+        const auto packindices_sub = Kokkos::create_mirror_view(interf.packindices_sub);
+        const auto packindices_schur = Kokkos::create_mirror_view(interf.packindices_schur);
+
+
+        // Fill packindices_sub and packindices_schur
+        for (local_ordinal_type local_sub_ip=0; local_sub_ip<n_subparts_per_part;++local_sub_ip) {
+          for (local_ordinal_type local_pack_ip=0; local_pack_ip<npacks_per_subpart;++local_pack_ip) {
+            packindices_sub(local_sub_ip * npacks_per_subpart + local_pack_ip) = 2 * local_sub_ip * npacks_per_subpart + local_pack_ip;
+            if (local_sub_ip != n_subparts_per_part-1) {
```
[the remark this answers is not in the document]
Comment (kliegeois) [reply packages/ifpack2/src/Ifpack2_BlockTriDiContainer_impl.hpp:1290]: In the code above, we reduce the number of lines of code but I am fine either ways.

**Review by lucbv — DISMISSED** [review]
lgtm

**Comment by lucbv** [comment]
Need a rebase to get the Experimental::HIP changes in

**Review by lucbv — APPROVED** [review]

**Comment by kliegeois** [comment]
I do not see a performance regression associated to this PR on V100 (actually I saw a small performance improvement even without using sub lines). Therefore, I am merging this PR.

**Comment by lucbv** [comment]
Thanks, this is good news!

**Pull request merged**